### PR TITLE
Chapter 0-4

### PR DESCRIPTION
**Pull request by EnoiuDiana**

### CODE REVIEW

File: shop/myFlywayConfig.conf [hunk 1]
```diff
@@ -0,0 +1,5 @@
+flyway.user=sa
```
Comment (athinabarbul) [inline shop/myFlywayConfig.conf]: You should define your flyway configuration in the application.properties file. In this way, the migration is done automatically at the application startup.

File: shop/src/main/java/ro/msg/learning/shop/controller/OrderController.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+@RestController
+@RequestMapping("/order")
+public class OrderController {
+
+    private final OrderService orderService;
+    private final OrderDetailMapper orderDetailMapper;
+    private final PlacedOrderMapper placedOrderMapper;
+
+
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/controller/OrderController.java]: One white line here is sufficient

File: shop/src/main/java/ro/msg/learning/shop/controller/OrderController.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+    public OrderController(OrderService orderService, OrderDetailMapper orderDetailMapper, PlacedOrderMapper placedOrderMapper) {
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/controller/OrderController.java]: You should use here too the @RequiredArgsConstructor annotation from Lombok

File: shop/src/main/java/ro/msg/learning/shop/controller/OrderController.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        this.orderService = orderService;
+        this.orderDetailMapper = orderDetailMapper;
+        this.placedOrderMapper = placedOrderMapper;
+    }
+
+    @PostMapping(value = "/create_order")
+    public ResponseEntity<PlacedOrderDTO> createOrder(@RequestBody CreateOrderDTO createOrderDTO) {
+/*        productService.createProduct(productMapper.convertFromCreationDTO(createProductDTO),
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/controller/OrderController.java]: Don't push commented code

File: shop/src/main/java/ro/msg/learning/shop/controller/OrderController.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import ro.msg.learning.shop.controller.mapper.OrderDetailMapper;
+import ro.msg.learning.shop.controller.mapper.PlacedOrderMapper;
+import ro.msg.learning.shop.dto.CreateOrderDTO;
+import ro.msg.learning.shop.dto.CreateProductDTO;
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/controller/OrderController.java]: Be aware of unused imports.

File: shop/src/main/java/ro/msg/learning/shop/controller/ProductController.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+    @PostMapping(value = "/create_product")
+    public ResponseEntity<Object> createProduct(@RequestBody CreateProductDTO createProductDTO) {
+        productService.createProduct(productMapper.convertFromCreationDTO(createProductDTO),
+                createProductDTO.getProductCategoryId(), createProductDTO.getSupplierId());
+        return new ResponseEntity<>("Product is created successfully", HttpStatus.CREATED);
+    }
+
+    @DeleteMapping(value = "/delete_product/{id}")
+    public ResponseEntity<Object> delete(@PathVariable("id") int id) {
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/controller/ProductController.java]: If the method parameter and the PathVariable have the same name then it's not necessary to set the value of the PathVariable specifically.

File: shop/src/main/java/ro/msg/learning/shop/controller/ProductController.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+@RequestMapping("/product")
+public class ProductController {
+
+    private final ProductService productService;
+    private final ProductCategoryService productCategoryService;
+
+    private final ProductMapper productMapper;
+
+    @Autowired
+    public ProductController(ProductService productService, ProductCategoryService productCategoryService, ProductMapper productMapper) {
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/controller/ProductController.java]: Same as before, you can also use the annotation from Lombok

File: shop/src/main/java/ro/msg/learning/shop/controller/ProductController.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+@RestController
+@RequestMapping("/product")
+public class ProductController {
+
+    private final ProductService productService;
+    private final ProductCategoryService productCategoryService;
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/controller/ProductController.java]: Try to be consistent here. Either you concatenate all the declarations or use a white line between them.

File: shop/src/main/java/ro/msg/learning/shop/controller/mapper/ProductCategoryMapper.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
+package ro.msg.learning.shop.controller.mapper;
+
+import org.springframework.stereotype.Component;
+import ro.msg.learning.shop.dto.ProductCategoryDTO;
+import ro.msg.learning.shop.model.ProductCategory;
+
+@Component
+public class ProductCategoryMapper {
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/controller/mapper/ProductCategoryMapper.java:8]: I really like the idea that you did separate mapper classes for each entity. Good job

File: shop/src/main/java/ro/msg/learning/shop/dto/CreateProductDTO.java [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+import java.math.BigDecimal;
+
+@Data
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/dto/CreateProductDTO.java]: No need for a white line

File: shop/src/main/java/ro/msg/learning/shop/dto/ProductDTO.java [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+import ro.msg.learning.shop.model.ProductCategory;
+import ro.msg.learning.shop.model.Supplier;
+
+import java.math.BigDecimal;
+
+@Data
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/dto/ProductDTO.java]: Same here. No need for a white line

File: shop/src/main/java/ro/msg/learning/shop/model/OrderDetail.java [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    private PlacedOrder placedOrder;
+
+    @ManyToOne
+    @JoinColumn(name = "product_id", nullable = false)
+    private Product product;
+
+    @Column(nullable = false)
+    private int quantity;
+
+/*    public OrderDetail(Product foundProduct, int quantity) {
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/model/OrderDetail.java]: Don't push commented code

File: shop/src/main/java/ro/msg/learning/shop/service/OrderService.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+    private final LocationRepository locationRepository;
+
+    private final ProductRepository productRepository;
+
+    private final LocationStrategy locationStrategy;
+
+    private final StockRepository stockRepository;
+
+
+    public OrderService(PlacedOrderRepository placedOrderRepository, CustomerRepository customerRepository, LocationRepository locationRepository, ProductRepository productRepository, LocationStrategy locationStrategy, StockRepository stockRepository) {
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/service/OrderService.java]: @RequiredArgsConstructor could be helpful here.

File: shop/src/main/java/ro/msg/learning/shop/service/OrderService.java [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+        this.placedOrderRepository = placedOrderRepository;
+        this.customerRepository = customerRepository;
+        this.locationRepository = locationRepository;
+        this.productRepository = productRepository;
+        this.locationStrategy = locationStrategy;
+        this.stockRepository = stockRepository;
+    }
+
+    private Set<OrderDetail> findProductsForOrderDetail(PlacedOrder placeOrder, List<OrderDetailProductIdAndQuantity> productIdAndQuantityList) {
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/service/OrderService.java]: Private methods should be placed after all the public methods.

File: shop/src/main/java/ro/msg/learning/shop/service/ProductService.java [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+import java.util.stream.Collectors;
+
+@Service
+public class ProductService {
+
+    private final ProductRepository productRepository;
+    private final ProductCategoryRepository productCategoryRepository;
+    private final SupplierRepository supplierRepository;
+
+
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/service/ProductService.java:19]: One white line is sufficient here

File: shop/src/main/java/ro/msg/learning/shop/service/ProductService.java [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+    }
+
+    public void updateProduct(Product product, int productCategoryId, int supplierId) {
+        if(productRepository.existsById(product.getId())) {
+            saveProduct(product, productCategoryId, supplierId);
+        } else {
+            throw new NotFoundException("Product not found");
+        }
+    }
+
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/service/ProductService.java:56]: Same as above. One white line is sufficient.

File: shop/src/main/java/ro/msg/learning/shop/service/OrderService.java [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+    }
+
+    public PlacedOrder placeOrder(PlacedOrder placedOrder, int customerID, List<OrderDetailProductIdAndQuantity> productIdAndQuantityList) {
+        if(customerRepository.existsById(customerID)) {
+            Customer customer = customerRepository.getReferenceById(customerID);
+            placedOrder.setCustomer(customer);
+            Set<OrderDetail> orderDetails = findProductsForOrderDetail(placedOrder, productIdAndQuantityList);
+            placedOrder.setOrderDetails(orderDetails);
+
+            //implement strategy location
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/service/OrderService.java]: You can remove this type of comment. These are only noise comments that restate the obvious.

File: shop/src/main/java/ro/msg/learning/shop/service/OrderService.java [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+            List<Stock> foundStocks = locationStrategy.getStockLocations(productIdAndQuantityList);
+            Location location = locationRepository.getReferenceById(foundStocks.get(0).getLocation().getId());
+            placedOrder.setLocation(location);
+
+            //place order
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/service/OrderService.java]: Same as above

File: shop/src/main/java/ro/msg/learning/shop/service/ProductService.java [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+        }
+        else throw new NotFoundException("product not found");
+    }
+
+    public void createProduct(Product product, int productCategoryId, int supplierId) {
+        saveProduct(product, productCategoryId, supplierId);
+
+    }
+
+    private void saveProduct(Product product, int productCategoryId, int supplierId) {
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/service/ProductService.java]: You should place the private methods after the public methods.

File: shop/src/main/java/ro/msg/learning/shop/service/OrderService.java [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+@Service
+public class OrderService {
+
+    private final PlacedOrderRepository placedOrderRepository;
+
+    private final CustomerRepository customerRepository;
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/service/OrderService.java]: The service from one entity should contain only its own repository. You should create for each entity that you use a different service with its own methods and use the service here.

File: shop/src/main/java/ro/msg/learning/shop/service/OrderService.java [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+    public OrderService(PlacedOrderRepository placedOrderRepository, CustomerRepository customerRepository, LocationRepository locationRepository, ProductRepository productRepository, LocationStrategy locationStrategy, StockRepository stockRepository) {
+        this.placedOrderRepository = placedOrderRepository;
+        this.customerRepository = customerRepository;
+        this.locationRepository = locationRepository;
+        this.productRepository = productRepository;
+        this.locationStrategy = locationStrategy;
+        this.stockRepository = stockRepository;
+    }
+
+    private Set<OrderDetail> findProductsForOrderDetail(PlacedOrder placeOrder, List<OrderDetail> productIdAndQuantityList) {
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/service/OrderService.java:48]: Declare your private methods after the public methods.

File: shop/src/main/java/ro/msg/learning/shop/service/strategy/SingleLocationStrategy.java [hunk 22]
```diff
@@ -0,0 +1,48 @@
...
+import ro.msg.learning.shop.repository.LocationRepository;
+import ro.msg.learning.shop.repository.StockRepository;
+import ro.msg.learning.shop.service.exceptions.NotFoundException;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class SingleLocationStrategy implements LocationStrategy{
+
+    private final StockRepository stockRepository;
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/service/strategy/SingleLocationStrategy.java]: You should inject at this level the services and not the repositories of the entities. The only layer that is communicating directly with the repositories is the service layer. Same applies for the other strategy implementation

File: shop/src/main/java/ro/msg/learning/shop/config/StrategyConfig.java [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+
+    private final StockRepository stockRepository;
+    private final LocationRepository locationRepository;
+
+    @Value("${strategy}")
+    private int strategy;
+
+    @Bean
+    public LocationStrategy getStrategyType() {
+        if(strategy == 0) {
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/config/StrategyConfig.java]: Try using enums for the static values of your strategies. Basically, we would have an enum containing the two values for the strategies (eg. enum LocationStrategy with values SINGLE_LOCATION_STRATEGY and MOST_ABUNDANT_STRATEGY).

File: shop/src/main/resources/db/migration/V10__first_insert_migration.sql [hunk 24]
```diff
@@ -0,0 +1,3 @@
+INSERT INTO PRODUCTCATEGORY(NAME, DESCRIPTION) VALUES ( 'laptop', 'this is laptop category' );
```
Comment (athinabarbul) [inline shop/src/main/resources/db/migration/V10__first_insert_migration.sql]: You should create a specific migration file for the main operations here (eg. V1 - table creation, V2-populate tables etc). You have too many major version migration files. Imagine having double the number of tables you have now.

**Review by athinabarbul — CHANGES_REQUESTED** [review]

**Comment by EnoiuDiana** [comment]
I have made all the necessary changes, you can check again now if everything is ok :)

File: shop/src/main/java/ro/msg/learning/shop/config/StrategyConfig.java [hunk 25]
```diff
@@ -0,0 +1,38 @@
...
+import ro.msg.learning.shop.service.StockService;
+import ro.msg.learning.shop.service.strategy.LocationStrategy;
+import ro.msg.learning.shop.service.strategy.MultipleLocationsStrategy;
+import ro.msg.learning.shop.service.strategy.SingleLocationStrategy;
+
+@Configuration
+@RequiredArgsConstructor
+public class StrategyConfig {
+
+    public enum StrategySelected {
```
Comment (athinabarbul) [inline shop/src/main/java/ro/msg/learning/shop/config/StrategyConfig.java]: You can move this enum into a separate file.

**Review by athinabarbul — APPROVED** [review]
One simple change and then you can merge the PR. Nicely done :)

**Pull request merged**